src/session.c

```diff
@@ -178,56 +180,129 @@ static P11PROV_SESSION *p11prov_session_ref(P11PROV_SESSION *session)
     return NULL;
 }
 
-static CK_RV p11prov_session_open(P11PROV_SESSION *session, bool login,
-                                  CK_UTF8CHAR_PTR pin, CK_ULONG pinlen)
+/* Calculates the start time and then nano-sleeps by 'interval' time.
+ * On the first invocation the content of start_time must be 0.
+ * The content of start_time must not be altered outside this function after
+ * the first invocation.
+ * This function does not guarantee each sleep is 'interval' long.
+ *
+ * Returns true if max_wait has not been reached yet.
+ * Returns false on an error or if max_wait is exceeded.
+ */
+static bool cyclewait_with_timeout(uint64_t max_wait, uint64_t interval,
+                                   uint64_t *start_time)
+{
+#define NANOS_SEC 1000000000
+    uint64_t current_time;
+    struct timespec ts;
+    int ret;
+
+    ret = clock_gettime(CLOCK_MONOTONIC, &ts);
+    if (ret != 0) {
+        return false;
+    }
+
+    current_time = ts.tv_sec * NANOS_SEC + ts.tv_nsec;
+    if (*start_time == 0) {
+        *start_time = current_time;
+    } else {
+        if (current_time > *start_time + max_wait) {
+            return false;
+        }
+    }
+
+    ts.tv_sec = interval / NANOS_SEC;
+    ts.tv_nsec = interval % NANOS_SEC;
+    ret = nanosleep(&ts, NULL);
+    if (ret != 0 && ret != EINTR) {
+        return false;
+    }
+
+    return true;
+}
+
+/* in nanoseconds, 1 seconds */
+#define MAX_WAIT 1000000000
+/* sleep interval, 50 microseconds (max 20 attempts) */
+#define SLEEP 50000
+static CK_RV internal_session_open(P11PROV_SESSION_POOL *p, CK_SLOT_ID slot,
+                                   CK_SESSION_HANDLE *session)
 {
-    P11PROV_SESSION_POOL *p = session->pool;
-    CK_ULONG cur_sessions;
     CK_FUNCTION_LIST *f;
+    bool wait_ok = true;
+    CK_ULONG cs = 0;
+    uint64_t startime = 0;
     CK_RV ret;
 
-    f = p11prov_ctx_fns(session->provctx);
+    f = p11prov_ctx_fns(p->provctx);
     if (f == NULL) {
         return CKR_FUNCTION_FAILED;
     }
 
-    cur_sessions = __atomic_add_fetch(&p->cur_sessions, 1, __ATOMIC_SEQ_CST);
-    if (cur_sessions > p->max_sessions) {
-        ret = CKR_SESSION_COUNT;
-        P11PROV_raise(session->provctx, ret, "Max sessions limit");
-        goto done;
+    while (wait_ok) {
+        if (cs == 0) {
+            cs = __atomic_add_fetch(&p->cur_sessions, 1, __ATOMIC_SEQ_CST);
+            if (cs > p->max_sessions) {
+                (void)__atomic_sub_fetch(&p->cur_sessions, 1, __ATOMIC_SEQ_CST);
+                ret = CKR_SESSION_COUNT;
+                cs = 0;
+                wait_ok = cyclewait_with_timeout(MAX_WAIT, SLEEP, &startime);
+                continue;
+            }
+        }
```

Comment:
Doing custom semaphores with sleep polling sounds prone to race conditions and hard to validate. It also means there's a level of busy polling with the process waking up a lot. So sleeping with wakup sounds better fit here.

Maybe consider using `pthread_cond_wait` and `pthread_cond_signal` together with mutex.
Alternatively this could be implemented as semaphore with `sem_post` and `sem_timedwait`.

Reply:
I considered using pthread_cond_wait() and similar, but that only works when the application itself is the cause of the issue.
If the problem is a separate process temporarily keeping sessions busy, we wouldn't be able to handle it.

The problem with sem_post/sem_timedwait is that you need a semaphore, but cur_sessions is a counter with a variable maximum. I would have to revers cur_sessions to be something like cur_free_sessions, then perhaps we could use it as a semaphore ... I will look into it after I address the basic issues in this review.

I think it is ok to busy-poll in this case anyway.

Reply:
So after some more thinking and seeing if I can use sem_post and sem_timedwait I realized that sem_timedwait does not let you chose a monotonic clock, which makes it vulnerable to clock changes.
I think I will stick with my crude implementation for now, given there are tests that prove it works.